crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2434,19 +2435,33 @@ impl<'db> TypeInferenceBuilder<'db> {
             op,
             values,
         } = bool_op;
+        Self::infer_chained_boolean_types(
+            self.db,
+            *op,
+            values.iter().map(|value| self.infer_expression(value)),
+            values.len(),
+        )
+    }
+
+    /// Computes the output of a chain of (one) boolean operation, consuming as input an iterator
+    /// of types. The iterator is consumed even if the boolean evaluation can be short-circuited,
+    /// in order to ensure the invariant that all expressions are evaluated when inferring types.
+    fn infer_chained_boolean_types<T: Into<Type<'db>>>(
+        db: &'db dyn Db,
+        op: ast::BoolOp,
+        values: impl IntoIterator<Item = T>,
+        n_values: usize,
+    ) -> Type<'db> {
         let mut done = false;
         UnionType::from_elements(
-            self.db,
-            values.iter().enumerate().map(|(i, value)| {
-                // We need to infer the type of every expression (that's an invariant maintained by
-                // type inference), even if we can short-circuit boolean evaluation of some of
-                // those types.
-                let value_ty = self.infer_expression(value);
+            db,
+            values.into_iter().enumerate().map(|(i, ty)| {
                 if done {
                     Type::Never
                 } else {
-                    let is_last = i == values.len() - 1;
-                    match (value_ty.bool(self.db), is_last, op) {
+                    let is_last = i == n_values - 1;
+                    let value_ty: Type<'db> = ty.into();
+                    match (value_ty.bool(db), is_last, op) {
                         (Truthiness::Ambiguous, _, _) => value_ty,
```

Comment:
See my comment on the last test, but maybe we could fix this by having a special case for `builtins.bool` (which will be a very common type here)
```suggestion
                        (Truthiness::Ambiguous, false, ast::BoolOp::And) => match value_ty {
                            // Ambiguous types that are not the last in the `and` chain can only be
                            // returned if they are falsy. In the special case of `builtins.bool`,
                            // being falsy is `Literal[False]`.
                            // TODO: we could do this optimisation for other literal that have a
                            // single falsy value (`""`, `0`, ...?)
                            Type::Instance(class) => class
                                .is_stdlib_symbol(db, "builtins", "bool")
                                .then(|| Type::BooleanLiteral(false))
                                .unwrap_or_else(|| value_ty),
                            _ => value_ty,
                        },
                        (Truthiness::Ambiguous, _, _) => value_ty,
```

Reply:
Let's do this as a follow-up, I created https://github.com/astral-sh/ruff/issues/13632 with some comments about it.